### PR TITLE
[benchmarks] Add LogBenchmark when BatchProcessor is used

### PR DESCRIPTION
**Pull request by cijothomas**
Added a Log Benchmark when `BatchExportProcessor` is used instead of a normal NoOpProcessor. SDK [special cases](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Logs/LoggerProviderSdk.cs#L99-L101) our `BatchExportProcessor` (not 3rd party authored), to use a different  LogRecord pooling mechanism. 

This benchmark will show the perf difference, when using that pooling mechanism. (It is much higher than the ThreadLocal based pooling, only useable in sync exporters like etw)

Also renamed Dummy -> NoOp processor, as the processor is a no-op, with no ability to accept alternate behavior.

The following shows the cost when BatchingProcessor is used (compared to other processor), which has to copy all the Log state+scope.

```
| OneProcessor                  | 111.558 ns | 2.9821 ns |  8.5082 ns | 109.311 ns | 0.0063 |      - |      40 B |
| BatchProcessor                | 263.650 ns | 5.2908 ns | 14.1223 ns | 258.984 ns | 0.0200 | 0.0043 |     128 B |
```

### CODE REVIEW

File: test/Benchmarks/Logs/LogBenchmarks.cs [hunk 1]
```diff
@@ -7,23 +7,24 @@
...
+
+| Method                        | Mean       | Error     | StdDev     | Median     | Gen0   | Gen1   | Allocated |
+|------------------------------ |-----------:|----------:|-----------:|-----------:|-------:|-------:|----------:|
+| NoListenerStringInterpolation | 135.503 ns | 2.7458 ns |  4.5114 ns | 135.391 ns | 0.0114 |      - |      72 B |
+| NoListenerExtensionMethod     |  40.218 ns | 0.8249 ns |  2.2581 ns |  39.809 ns | 0.0102 |      - |      64 B |
+| NoListener                    |   1.930 ns | 0.0626 ns |  0.1264 ns |   1.889 ns |      - |      - |         - |
+| UnnecessaryIsEnabledCheck     |   1.531 ns | 0.0542 ns |  0.1267 ns |   1.518 ns |      - |      - |         - |
+| CreateLoggerRepeatedly        |  53.797 ns | 1.0927 ns |  1.7331 ns |  53.401 ns | 0.0038 |      - |      24 B |
+| OneProcessor                  | 111.558 ns | 2.9821 ns |  8.5082 ns | 109.311 ns | 0.0063 |      - |      40 B |
+| BatchProcessor                | 263.650 ns | 5.2908 ns | 14.1223 ns | 258.984 ns | 0.0200 | 0.0043 |     128 B |
```
Comment (cijothomas) [inline test/Benchmarks/Logs/LogBenchmarks.cs:25]: @CodeBlanch What might be the reason why there is a notable difference when the `LogRecord` comes from a heap pool vs thread-local? There is no extra copy involved...

Comment (utpilla) [reply test/Benchmarks/Logs/LogBenchmarks.cs:25]: Do we just want to measure the perf difference specific to the pooling mechanism? If yes, then the benchmarks should just compare `LogRecordThreadStaticPool.Instance.Rent()`  vs `LogRecordSharedPool.Current.Rent()`.

Testing this using `BatchProcessor` would add all the `BatchProcessor` related overhead as well such as buffering the log record attributes, adding data to `CircularBuffer` etc.

Comment (cijothomas) [reply test/Benchmarks/Logs/LogBenchmarks.cs:25]: I want to measure the overall cost when using Batch vs Reentrant processors.

Anyway, your comment can explain the perf difference as SharedPool requires us to copy the attributes for safe processing later, while ThreadLocalPool does not need to :)

Comment (utpilla) [reply test/Benchmarks/Logs/LogBenchmarks.cs:25]: > SharedPool requires us to copy the attributes for safe processing later

You mean `BatchProcessor` requires us to copy the attributes for safe processing later?

Comment (cijothomas) [reply test/Benchmarks/Logs/LogBenchmarks.cs:25]: > > SharedPool requires us to copy the attributes for safe processing later
> 
> You mean `BatchProcessor` requires us to copy the attributes for safe processing later?

Yes.

**Review by Kielek — APPROVED** [review]

File: test/Benchmarks/Logs/LogBenchmarks.cs [hunk 2]
```diff
@@ -134,7 +150,15 @@ public void ThreeProcessors()
         this.loggerWithThreeProcessors.SayHello(FoodName, FoodPrice);
     }
 
-    internal class DummyLogProcessor : BaseProcessor<LogRecord>
+    internal class NoOpLogProcessor : BaseProcessor<LogRecord>
+    {
+    }
+
+    internal class NoOpExporter : BaseExporter<LogRecord>
```
Comment (Kielek) [inline test/Benchmarks/Logs/LogBenchmarks.cs:157]: nit:

```suggestion
    internal class NoOpLogProcessor : BaseProcessor<LogRecord>
    {
    }

    internal class NoOpLogExporter : BaseExporter<LogRecord>
```

or

```suggestion
    internal class NoOpProcessor : BaseProcessor<LogRecord>
    {
    }

    internal class NoOpExporter : BaseExporter<LogRecord>
```

File: test/Benchmarks/Logs/LogBenchmarks.cs [hunk 3]
```diff
@@ -134,7 +150,15 @@ public void ThreeProcessors()
         this.loggerWithThreeProcessors.SayHello(FoodName, FoodPrice);
     }
 
-    internal class DummyLogProcessor : BaseProcessor<LogRecord>
+    internal class NoOpLogProcessor : BaseProcessor<LogRecord>
```
Comment (reyang) [inline test/Benchmarks/Logs/LogBenchmarks.cs:153]:
```suggestion
    internal class NoopLogProcessor : BaseProcessor<LogRecord>
```

**Review by rajkumar-rangaraj — APPROVED** [review]

File: test/Benchmarks/Logs/LogBenchmarks.cs (hunk 1 above)
Comment (utpilla) [reply test/Benchmarks/Logs/LogBenchmarks.cs:25]: You could update the PR description as well. It mainly attributes thread pooling differences to make a case for adding a `BatchProcessor` benchmark.

File: test/Benchmarks/Logs/LogBenchmarks.cs [hunk 4]
```diff
@@ -134,7 +150,15 @@ public void ThreeProcessors()
         this.loggerWithThreeProcessors.SayHello(FoodName, FoodPrice);
     }
 
-    internal class DummyLogProcessor : BaseProcessor<LogRecord>
+    internal class NoopLogProcessor : BaseProcessor<LogRecord>
```
Comment (cijothomas) [inline test/Benchmarks/Logs/LogBenchmarks.cs:153]:
```suggestion
    internal class NoOpLogProcessor: BaseProcessor<LogRecord>
```

File: test/Benchmarks/Logs/LogBenchmarks.cs (hunk 3 above)
Comment (reyang) [reply test/Benchmarks/Logs/LogBenchmarks.cs:153]: It this updated? @cijothomas

**Pull request merged**